packages/@uppy/dashboard/src/Dashboard.tsx

```diff
@@ -9,124 +19,274 @@ import { defaultPickerIcon } from '@uppy/provider-views'
 
 import { nanoid } from 'nanoid/non-secure'
 import memoizeOne from 'memoize-one'
-import * as trapFocus from './utils/trapFocus.js'
-import createSuperFocus from './utils/createSuperFocus.js'
-import DashboardUI from './components/Dashboard.jsx'
+import * as trapFocus from './utils/trapFocus.ts'
+import createSuperFocus from './utils/createSuperFocus.ts'
+import DashboardUI from './components/Dashboard.tsx'
 
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore We don't want TS to generate types for the package.json
 import packageJson from '../package.json'
-import locale from './locale.js'
+import locale from './locale.ts'
+
+type GenericEventCallback = () => void
+export type DashboardFileEditStartCallback<M extends Meta, B extends Body> = (
+  file?: UppyFile<M, B>,
+) => void
+export type DashboardFileEditCompleteCallback<
+  M extends Meta,
+  B extends Body,
+> = (file?: UppyFile<M, B>) => void
+export type DashboardShowPlanelCallback = (id: string) => void
+declare module '@uppy/core' {
+  export interface UppyEventMap<M extends Meta, B extends Body> {
+    'dashboard:modal-open': GenericEventCallback
+    'dashboard:modal-closed': GenericEventCallback
+    'dashboard:show-panel': DashboardShowPlanelCallback
+    'dashboard:file-edit-start': DashboardFileEditStartCallback<M, B>
+    'dashboard:file-edit-complete': DashboardFileEditCompleteCallback<M, B>
+    'dashboard:close-panel': (id: string | undefined) => void
+    'restore-canceled': GenericEventCallback
+  }
+}
+
+interface PromiseWithResolvers<T> {
+  promise: Promise<T>
+  resolve: (value: T | PromiseLike<T>) => void
+  reject: (reason?: any) => void
+}
 
-const memoize = memoizeOne.default || memoizeOne
+const memoize = ((memoizeOne as any).default as false) || memoizeOne
 
 const TAB_KEY = 9
 const ESC_KEY = 27
 
-function createPromise () {
-  const o = {}
-  o.promise = new Promise((resolve, reject) => {
+function createPromise<T>(): PromiseWithResolvers<T> {
+  const o = {} as PromiseWithResolvers<T>
+  o.promise = new Promise<T>((resolve, reject) => {
     o.resolve = resolve
     o.reject = reject
   })
   return o
 }
 
+type FieldRenderOptions = {
+  value: string
+  onChange: (newVal: string) => void
+  fieldCSSClasses: { text: string }
+  required: boolean
+  form: string
+}
+
+type PreactRender = (
+  node: any,
+  params: Record<string, unknown> | null,
+  ...children: any[]
+) => any
+
+interface MetaField {
+  id: string
+  name: string
+  placeholder?: string
+  render?: (field: FieldRenderOptions, h: PreactRender) => any
```

Comment:
Since this is user facing, can we be more specific than `any` for the return value?

Reply:
I simply copy/pasted the existing type: https://github.com/transloadit/uppy/blob/83e252781017afa508cd5229bae1e95f405e8a1e/packages/@uppy/dashboard/types/index.d.ts#L26-L31

Reply:
I don't mind some `any`'s for when we can't work out internals, but user facing types should be sound. Can we improve this?

Reply:
We certainly can, I'd rather not be the one doing it though. I think the improvement can (and should) happen independently from this PR.

Reply:
It's just changing one return type? Am I misunderstanding something or why so reluctant?

Reply:
Someone needs to do the work of figuring out what this used for, and what type we should put instead of `any`; as I said, I'd rather not be the one doing that work.

Reply:
You submit a PR with +1,992 −1,142 changes, I'm asking you for _one_ change in an important user facing type, while normally a PR this size has dozens of comments. You are the expert on this types because you wrote them, forcing someone else to contextualize what you already did sounds more unfair to me than letting the author of the PR resolve a single request.

Reply:
> You are the expert on this types because you wrote them

I did not write this type, according to `git blame`, Renée did.



> forcing someone else to contextualize what you already did sounds more unfair to me than letting the author of the PR resolve a single request

Sorry if that's how I came across, but AFAICT I'm in no way in a better position than someone else to figure this out.

> You submit a PR with +1,992 −1,142 changes

That's precisely why I'd rather not be the one doing that work.

Reply:
1e733c1 took two minutes. I hope this serves as an inspiration that it's sometimes better to just take a look than argue.